CSIS: Retry mechanism for notifications

fredrikdanebjer opened the pull request.

fredrikdanebjer committed:
Bluetooth: Audio: Change CSIP notification to use Atomic flags

Reformated csip_set_member to use atomic flags for notifications
instead of a pending notification array. This better conforms the
style of the code to how its done in other LE Audio services.

Signed-off-by: Fredrik Danebjer <frdn@demant.com>

fredrikdanebjer committed:
Bluetooth: Audio: Add resend mechanic to CSIP notifications

Added retry mechanic to CSIP notifications by adding them into system
workqueue. Bookkeeping is done for notifications for all clients, and
potentially disconncted clients will also receive notifications on
reconnect. This also adds mechanic to restore the local clients list
upon registration of the service, as well as BSIMs to test notify on
reconnect.

Signed-off-by: Fredrik Danebjer <frdn@demant.com>

kartben commented:
FWIW doc failure is fixed (#62967) -- you may please rebase. Sorry for the inconvenience!

Thalley commented on subsys/bluetooth/audio/csip_set_coordinator.c:
```diff
@@ -1575,8 +1575,8 @@ static int bt_csip_set_coordinator_get_lock_state(
 		 */
-
-		if (!active.oap_cb(active.info, active.members, active.members_count)) {
+		if (!active.oap_cb ||
```
```suggestion
		if (active.oap_cb == NULL ||
```

I prefer the explicit NULL checks

Thalley replied:
I think you forgot this one :)

Thalley commented on subsys/bluetooth/audio/csip_set_member.c:
```diff
@@ -39,6 +39,27 @@
+ * the array containing this struct, if the security manager overwrites
+ * the oldest keys, we also overwrite the oldest entry
+ */
+#if defined(CONFIG_BT_KEYS_OVERWRITE_OLDEST)
+	uint32_t age;
+#endif /* CONFIG_BT_KEYS_OVERWRITE_OLDEST */
```
Since you are refactoring this part, you can consider squeezing in the fix for https://github.com/zephyrproject-rtos/zephyr/issues/61763 as well, but up to you

fredrikdanebjer replied:
Ill take a look at this as well

Thalley commented on subsys/bluetooth/audio/csip_set_member.c:
```diff
@@ -74,13 +101,65 @@ static bool is_last_client_to_write(const struct bt_csip_set_member_svc_inst *sv
 }
 
-static void notify_lock_value(const struct bt_csip_set_member_svc_inst *svc_inst,
+void csip_gatt_notify_complete_cb(struct bt_conn *conn, void *user_data)
```
```suggestion
static void csip_gatt_notify_complete_cb(struct bt_conn *conn, void *user_data)
```

Thalley commented on subsys/bluetooth/audio/csip_set_member.c:
```diff
@@ -74,13 +101,65 @@ static bool is_last_client_to_write(const struct bt_csip_set_member_svc_inst *sv
+	k_work_submit(&deferred_nfy_work);
+}
+
+static int csip_gatt_notify(struct bt_conn *conn,
+			    const struct bt_uuid *uuid,
+			    const struct bt_gatt_attr *attr,
+			    const void *data,
+			    uint16_t len)
```
```suggestion
static int csip_gatt_notify_set_lock(struct bt_conn *conn,
			    const struct bt_gatt_attr *attr,
			    const void *data,
			    uint16_t len)
```

Since this is only used for the set lock, we can omit the UUID from the parameters (and actually also the `len`) since these are constant values

Thalley commented on subsys/bluetooth/audio/csip_set_member.c:
```diff
@@ -74,13 +101,65 @@ static bool is_last_client_to_write(const struct bt_csip_set_member_svc_inst *sv
-			    &svc_inst->set_lock,
-			    sizeof(svc_inst->set_lock));
+	LOG_DBG("");
+	return csip_gatt_notify(conn, BT_UUID_CSIS_SET_LOCK,
+			 svc_inst->service_p->attrs,
+			 &svc_inst->set_lock,
+			 sizeof(svc_inst->set_lock));
```
Indentation

Thalley commented on subsys/bluetooth/audio/csip_set_member.c:
```diff
@@ -746,10 +833,72 @@ static void remove_csis_char(const struct bt_uuid *uuid, struct bt_gatt_service
+}
+
+static void add_bonded_addr_to_client_list(const struct bt_bond_info *info, void *data)
+{
+
+	for (int i = 0; i < ARRAY_SIZE(svc_insts); i++) {
```
```suggestion
{

	for (size_t i = 0U; i < ARRAY_SIZE(svc_insts); i++) {
```

Thalley commented on subsys/bluetooth/audio/csip_set_member.c:
```diff
@@ -746,10 +833,72 @@ static void remove_csis_char(const struct bt_uuid *uuid, struct bt_gatt_service
+	for (int i = 0; i < ARRAY_SIZE(svc_insts); i++) {
+		struct bt_csip_set_member_svc_inst *svc_inst = &svc_insts[i];
+
+		for (int j = 1; j < ARRAY_SIZE(svc_inst->clients); i++) {
```
```suggestion
		for (size_t j = 1U; j < ARRAY_SIZE(svc_inst->clients); j++) {
```

Thalley commented on tests/bsim/bluetooth/audio/src/csip_notify_server_test.c:
```diff
@@ -0,0 +1,126 @@
+	if (!attr) {
+		printk("No BT_UUID_PACS_SNK attribute found\n");
+		return false;
+	}
+	return bt_gatt_is_subscribed(conn, attr, BT_GATT_CCC_NOTIFY);
```
```suggestion
	}
	
	return bt_gatt_is_subscribed(conn, attr, BT_GATT_CCC_NOTIFY);
```

Thalley commented on tests/bsim/bluetooth/audio/src/csip_notify_server_test.c:
```diff
@@ -0,0 +1,126 @@
+	}
+
+	printk("Registering CSIP Set Member\n");
+	struct bt_csip_set_member_register_param csip_params = {
+		.set_size = 1,
+		.rank = 1,
+		.lockable = true,
+		.cb       = &csip_cb,
+	};
```
Please declare variables at top of scope

Thalley commented on tests/bsim/bluetooth/audio/src/csip_notify_server_test.c:
```diff
@@ -0,0 +1,126 @@
+	}
+	printk("Subscribed\n");
+
+	err = bt_csip_set_member_lock(svc_inst, true, false);
+
```
Missing error check

Thalley commented on tests/bsim/bluetooth/audio/src/csip_notify_server_test.c:
```diff
@@ -0,0 +1,126 @@
+	}
+
+	/* Trigger changes while device is disconnected */
+	err = bt_csip_set_member_lock(svc_inst, false, false);
+
```
Missing error check

fredrikdanebjer commented on tests/bsim/bluetooth/audio/src/csip_notify_client_test.c:
```diff
@@ -0,0 +1,198 @@
+		subscribe_params = &csip_instance.csip_set_lock_sub;
+		subscribe_params->notify = csip_notify_handler;
+		subscribe_params->value = BT_GATT_CCC_NOTIFY;
+		subscribe_params->ccc_handle = attr->handle;
+
+		err = bt_gatt_subscribe(conn, subscribe_params);
+		if (err && err != -EALREADY) {
+			printk("Subscribe failed (err %d)\n", err);
+		} else {
+			SET_FLAG(flag_csip_set_lock_discovered);
+			printk("[SUBSCRIBED]\n");
+		}
+	} else {
+		printk("Unknown handle at %d\n", attr->handle);
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	return BT_GATT_ITER_STOP;
+}
+
+static void discover_and_subscribe_csip_lock(void)
+{
+	int err = 0;
+
+	printk("%s\n", __func__);
+	printk("									CSIP\n");
+
+	memcpy(&uuid, BT_UUID_CSIS_SET_LOCK, sizeof(uuid));
+	csip_instance.discover_params.uuid = &uuid.uuid;
+	csip_instance.discover_params.start_handle = BT_ATT_FIRST_ATTRIBUTE_HANDLE;
+	csip_instance.discover_params.end_handle = BT_ATT_LAST_ATTRIBUTE_HANDLE;
+	csip_instance.discover_params.type = BT_GATT_DISCOVER_CHARACTERISTIC;
+	csip_instance.discover_params.func = discover_csip_set_lock;
+
+	err = bt_gatt_discover(default_conn, &csip_instance.discover_params);
+	if (err != 0) {
+		FAIL("Service Discovery failed (err %d)\n", err);
+		return;
+	}
+}
```
I assume that you refer to using f.x. `bt_csip_set_coordinator_discover` in the `csip_set_coordinator.c` ?
If I would do that, then I would not have access setting `SET_FLAG(flag_csip_set_lock_discovered);`.

Thalley replied:
In the discover callback you get the `struct bt_csip_set_coordinator_set_member *member`, which contains one or more `struct bt_csip_set_coordinator_csis_inst` which has the `struct bt_csip_set_coordinator_set_info` which then has the `bool lockable`, which you can use to determine whether or not the set lock characteristic has been found. 

If there is something missing from the implementation, it should be fixed in the implementation, rather than creating a new implementation :)

fredrikdanebjer replied:
Thats fair enough. I wasn't thinking of this being something missing from implementation, since the test actually doesn't test the client functionality, only that the server behaves as it should. There is ofc. more value in adding our client as well here if possible.

Ill make adjustment according to your comments.

Thalley commented on subsys/bluetooth/audio/csip_set_member.c:
```diff
@@ -74,13 +101,63 @@ static bool is_last_client_to_write(const struct bt_csip_set_member_svc_inst *sv
+	k_work_submit(&deferred_nfy_work);
+}
+
+static int csip_gatt_notify_set_lock(struct bt_conn *conn,
+			    const struct bt_gatt_attr *attr,
+			    const void *data,
+			    uint16_t len)
```
Indentation

Thalley commented on subsys/bluetooth/audio/csip_set_member.c:
```diff
@@ -74,13 +101,63 @@ static bool is_last_client_to_write(const struct bt_csip_set_member_svc_inst *sv
+	if (err != 0) {
+		atomic_clear(&notify_rdy);
+
+		if (err == -ENOTCONN) {
+			return err;
+		}
```
Should this have been
```suggestion
		if (err != -ENOTCONN) {
			return err;
		}
```

fredrikdanebjer replied:
Yes :)

fredrikdanebjer committed:
Bluetooth: Audio: Make CSIP delete bonds in cache locally

This change from using the mechanic of CONFIG_BT_KEYS_OVERWRITE_OLDEST
to instead handle the bonding cache through bond_deleted.

Signed-off-by: Fredrik Danebjer <frdn@demant.com>

Thalley commented on tests/bsim/bluetooth/audio/src/csip_notify_client_test.c:
The majority of this and the server test already exist for the existing CSIP client and server test source files. Have you considered just appending the notify test as another test case in those files?

fredrikdanebjer replied:
I did yes, but I don't necessarily see the same function abstraction as with the other tests. Where the other tests just take a flag to condition the tests, but otherwise to the same procedure of connecting, discovering, performing ordered access procedure, etc, this one doesn't care about any of those steps, except the initial discovery. It only cares about disconnecting and getting the notification after reconnect. So adding it in the same test would kind of be:
1. Do same things until connected/discovered
2. Skip all the test procedures
3. Run own subroutine at the end for reconnection, rediscovery and notification, which no one else does

I'm sure you have a better understanding of how we actually should shape these tests, but this is at least my reasoning behind it. If we strictly want to keep all tests in the same profile/service in the same c file, then I guess your proposal makes most sense.

What do you think?

@larsgk ?

Thalley replied:
The reasoning behind my suggestion, is that at least 50% of the code you've added, already exist in the existing source file. 

We can keep this as is if you disagree, and then perhaps I can try to merge them later to see how much can really be used.

larsgk replied:
Can we have it as it is for now and potentially optimize/merge what's possible in a subsequent PR?

Thalley replied:
> Can we have it as it is for now and potentially optimize/merge what's possible in a subsequent PR?

Yes, @fredrikdanebjer believe there's not too much common between the multiple tests cases to put it in the same file, we can attempt that later

Thalley reviewed (approved):
The update of the test can be postponed for later
